### DISCUSSION TITLE
Does RainbowKit v2 EIP-6963 support extend to mobile?

### DISCUSSION THREAD

**Discussion by ablankstory** [rainbow-me/rainbowkit#1732] [Q&A]
Hi -

I noticed that when i go to this site (https://rainbowkit-example-git-rainbowkit-wagmi-v2-rainbowdotme.vercel.app/) on desktop and Connect, it recognizes MetaMask as an Installed (injected) Wallet. This is due to the support for EIP-6963, which is great! However, if I go to that same site in MetaMask's dapp browser on my iPhone, I don't get the injected wallet list. Rather, it just looks like the same modal.

So - will the "Installed" section also be available when connecting on mobile through an EIP-6963 supported wallet's dapp browser?

<img width="196" alt="Screenshot 2024-01-26 at 10 03 20 PM" src="https://github.com/rainbow-me/rainbowkit/assets/104391159/95fb7f06-cf9d-4126-b6a5-09617896eddd">

If so, we'll prioritize work on adding support for EIP-6963 right away.

Thanks!

**Comment by 0xmkh** [comment]
@ablankstory This is right we do not support EIP6963 on mobile. EIP6963 is not so well introduced on mobile that's why we've got only these wallets and they all have [mobile linking](https://docs.walletconnect.com/web3wallet/mobileLinking) with WalletConnect. 

I mean there is a few situation you might get EIP6963 in mobile through like metamask app browser, but this is not well known since you can't get injected connectors via mobile for most wallets and that's why we chose to go for the WalletConnect route instead without having the "Installed" section on mobile.

**Reply by ablankstory** [reply]
@KosmosKey  Thanks for the reply. Maybe if I explain my "issue", you may be able to provide some advice.

The Loopring Wallet has a dapp browser. We use trustweb3provider. When you go to a dapp and connect, users have to kind of "guess" which wallet to select:

- Browser Wallet
- Trust Wallet
- MetaMask

But it's not intuitive because these aren't "Loopring". So we get a lot of questions on which wallet to select. Additionally, WalletConnect doesn't work within the dapp browser - the prompt just kind of spins and nothing happens.

Since RainbowKit isn't accepting more wallets to add to the Custom Wallet List (which would help make the Loopring connector easily discoverable) - I was hoping we could rely on the EIP-6963 support to highlight Loopring Wallet in your modal.

Please let me know if anything here is unclear or if you'd like me to send a video to help explain the issue. 


Alternatively - is there a way for RainbowKit mobile to auto-detect browser wallet (when accessed in a dapp browser) and select it when connecting? Really trying to improve the UX for our users when accessing dapps within our dapp browser. Thanks!

**Reply by ablankstory** [reply]
I think ideally, for connecting to a dapp while in a DApp browser:

1. EIP-6963 support recognizes the wallet you are connecting from and shows the option first in the list (e.g., Loopring)

2. Add option/selection for "Browser Wallet" and place it first in the list. Similar to WalletConnect's web3modal- which works great from our dapp browser.

3. You consider continuing to allow wallets to request being added to your Custom Wallet List so dapp developers can easily find them.


WalletConnect has "Browser Wallet"
Family's ConnectKit has the injected wallet as an option.

Images of both while accessing their connectors via Loopring Wallet DApp Browser:
<img width="405" alt="Screenshot 2024-01-26 at 8 51 51 PM" src="https://github.com/rainbow-me/rainbowkit/assets/104391159/3282e465-6268-4a22-aa43-61e9ac801676">


Am I missing something obvious in RainbowKit? Thanks again for taking the time to read and respond.

**Reply by ablankstory** [reply]
Actually - this issue is kind of on our end. We use TrustWeb3Provider and RainbowKit thinks we're TrustWallet and hides "Browser Extension". 

I'm going to ask our team to look at our settings to see if we can adjust them so that we don't appear as TrustWallet.

Thanks.

**Reply by 0xmkh** [reply]
@ablankstory Sounds good. Meanwhile you're looking at that issue i've read your comments and it totally make sense. If loopring works in that way i would definitely have a look at mobile support either for the EIP6963 support or auto browser wallet detection.

Usually in mobile browsers the injected wallets aren't present and EIP6963 aren't detected also, but it seems like if you use dApp browser through metamask or loopring it'll support those options. 

That's why we went with the WalletConnect mobile linking route since not a lot of people are using dApp browsers through an mobile wallet app.

I'll get back to you on this and let you know if there's any updates on our end 🙏

**Reply by ablankstory** [reply]
@KosmosKey - hope you had a good weekend. I chatted a bit with WalletConnect and they actually support EIP-6963 in their web3modal. For example, if you go to web3modal.com and use MetaMask or Ronin, those wallets will jump to the top of the list because they are recognized via 6963.

We (Loopring) are going to work on adding support for 6963 in our wallet browser. It would be great if Rainbow could also support! Thanks!

**Reply by 0xmkh** [reply]
Sounds great. Will look into this and keep you updated 🙏

**Reply by ablankstory** [reply]
Also noting that I confirmed with Family/ConnectKit they also support EIP-6963 in wallet dapp browsers. I tested this using MM, Ronin, and OKX wallets.

Looking forward to having RainbowKit prioritize this! 🙏

**Reply by 0xmkh** [reply]
Great to know. We have it as a part of the roadmap. We're just prioritising getting the wagmi v2 out and then push more features on top of that. One of them is EIP6963 in wallet dapp browsers 👍

**Reply by ablankstory** [reply]
Thanks so much @KosmosKey. This is great to hear! 🙏🤝

**Reply by Volfor** [reply]
Hi @magiziz , any updates on this?
We have an sdk injecting wallet into a webpage and would love to see EIP6963 support for mobile browsers as well.

**Reply by 0xmkh** [reply]
Hey @Volfor. We're deprioritising this atm, but we'll give you a sound once it's ready

**Reply by vimtor** [reply]
We got Loopring wallet working doing the following:

```ts
import { connectorsForWallets, Wallet } from "@rainbow-me/rainbowkit";
import { injectedWallet } from "@rainbow-me/rainbowkit/wallets";

export const loopringWallet = (): Wallet => ({
  id: "loopring",
  name: "Loopring",
  iconUrl: "/landing/wallets/loopring.png",
  iconBackground: "#3774FF",
  downloadUrls: {
    android: "https://play.google.com/store/apps/details?id=loopring.defi.wallet",
    ios: "https://apps.apple.com/us/app/loopring-smart-wallet/id1550921126",
  },
  mobile: {
    getUri: (uri: string) => uri,
  },
  createConnector: injectedWallet().createConnector,
});
```

It would be nice if Rainbowkit exported the `getInjectedConnector` function.

**Reply by 0xmkh** [reply]
Hey. We can add this on our end and you can import `loopringWallet` through `@rainbow-me/rainbowkit/wallets`. How does that sound ?

**Reply by vimtor** [reply]
@magiziz that would be amazing!

**Comment by NORAVemp** [comment]
@magiziz  Does the mobile browser support the EIP6963 protocol?

**Reply by 0xmkh** [reply]
No not yet. There are some few bugs we spotted on some mobile browser like MetaMask. Hopefully should get fixed soon 👌
